Frameworks/CoreText/DWriteWrapper.mm

```diff
@@ -0,0 +1,392 @@
+
+    for (uint32_t i = 0; i < count; ++i) {
+        // Get the font family.
+        ComPtr<IDWriteFontFamily> fontFamily;
+        THROW_IF_FAILED(fontCollection->GetFontFamily(i, &fontFamily));
+
+        // Get a list of localized strings for the family name.
+        ComPtr<IDWriteLocalizedStrings> familyNames;
+        THROW_IF_FAILED(fontFamily->GetFamilyNames(&familyNames));
+
+        [fontFamilyNames addObject:_convertLocalizedStringToNSString(familyNames.Get())];
+    }
+
+    return [fontFamilyNames autorelease];
+}
+
+/**
+ * Helper method to create _CTLine objects given a _CTTypesetter, attributed string range to use and frame size to fit in.
+ *
+ * @parameter ts _CTTypesetter object to use.
+ * @parameter range attributed string range to use.
+ * @parameter frameSize size parameters of the frame to fit the text into.
+ *
+ * @return Unmutable array of _CTLine objects created with the requested parameters.
+ */
+static NSArray<_CTLine*>* _dwriteGetLines(_CTTypesetter* ts, CFRange range, CGRect frameSize) {
+    // Call custom renderer to get all glyph run details
+    _DWriteGlyphRunDetails glyphRunDetails;
+    _callCustomRender(ts, range, frameSize, &glyphRunDetails);
+
+    // Create _CTLine objects from the the obtained glyph run details
+    int numOfGlyphRuns = glyphRunDetails._glyphRuns.size();
+    int i = 0;
+    int j = 0;
+
+    NSMutableArray<_CTLine*>* lines = [NSMutableArray array];
+
+    while (j < numOfGlyphRuns) {
+        _CTLine* line = [[_CTLine new] autorelease];
+        NSMutableArray<_CTRun*>* runs = [NSMutableArray array];
+        uint32_t stringRange = 0;
+        // Glyph runs that have the same _baselineOriginY value are part of the the same Line.
+        while ((j < numOfGlyphRuns) && (glyphRunDetails._baselineOriginY[i] == glyphRunDetails._baselineOriginY[j])) {
+            j++;
+        }
+        while (i < j) {
+            // Create _CTRun objects and make them part of _CTLine
+            _CTRun* run = [[_CTRun new] autorelease];
+            run->_range.location = glyphRunDetails._glyphRunDescriptions[i]._textPosition;
+            run->_range.length = glyphRunDetails._glyphRunDescriptions[i]._stringLength;
+            run->_stringFragment = [ts->_string substringWithRange:NSMakeRangeFromCF(run->_range)];
+            run->_xPos = frameSize.origin.x + glyphRunDetails._baselineOriginX[i];
+            // TODO::
+            // This is a temp work around to get DWrite measurements to be usable with Cairo FreeType rasterizer. Remove this when
+            // we move away from Cairo Freetype. We need this workaround because DWrite coordinate system begins from top left but
+            // Cairo FreeType has its from bottom left.
+            run->_yPos = frameSize.origin.y + frameSize.size.height - glyphRunDetails._baselineOriginY[i];
+            run->_attributes = [ts->_attributedString attributesAtIndex:run->_range.location effectiveRange:NULL];
```

Comment:
<<see above comment>>

---
In reply to: [77757337](https://github.com/Microsoft/WinObjC/pull/950#discussion_r77757337) [](ancestors = 77757337)